### PR TITLE
Disallow certain non-filename characters in Export Text

### LINKED ISSUE #13664: Export Text when in Admin mode filename suggestion has invalid character ":"
### Windows Terminal version

1.14.1963.0

### Windows build number

22000.795

### Other Software

_No response_

### Steps to reproduce

Open Windows Terminal (Admin).
Have text in console window.
Right click on tab and choose "Export Text".

### Expected Behavior

Cleanse out the colon character and other invalid characters for filenames based on the tab name.

### Actual Behavior

Observe filename provided has "Admin: " as the leading text, and on Windows, a colon is not valid in filenames.

### PR DESCRIPTION
**Pull request by EliaSchiavon**
Added control in code for not allowed characters in the filename when saving a
tab.

Closes #13664

### CODE REVIEW

**Comment by EliaSchiavon** [comment]
@zadjii-msft not tried it yet, but something like this could work. Maybe it can be wrapped in a function `str_to_filename` and put somewhere else. 

Side-note: I'm a beginner in C++, so maybe there's a better way of implementing this (maybe using stl algorithms?). Let me know any feedback :)

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 1]
```diff
@@ -433,7 +433,13 @@ namespace winrt::TerminalApp::implementation
                         THROW_IF_FAILED(dialog->SetDefaultExtension(L"txt"));
 
                         // Default to using the tab title as the file name
-                        THROW_IF_FAILED(dialog->SetFileName((control.Title() + L".txt").c_str()));
+                        auto filename = { control.Title() };
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TabManagement.cpp:436]: `Title()` returns a `winrt::hstring` which is an immutable string. You first have to make it mutable by copying it into another container like `std::wstring`:
```cpp
std::wstring filename{ control.Title() };
```

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 2]
```diff
@@ -433,7 +433,13 @@ namespace winrt::TerminalApp::implementation
                         THROW_IF_FAILED(dialog->SetDefaultExtension(L"txt"));
 
                         // Default to using the tab title as the file name
-                        THROW_IF_FAILED(dialog->SetFileName((control.Title() + L".txt").c_str()));
+                        auto filename = { control.Title() };
+                        char unallowed_chars[] = "<>:\\/\"|?*";
+                        for (unsigned int i = 0; i < strlen(unallowed_chars); ++i)
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TabManagement.cpp:438]: You can iterate over strings more easily by putting it into a string-view:
```cpp
static constexpr std::wstring_view unallowedChars{ L"<>:\\/\"|?*" };
for (const auto& ch : unallowedChars) { ... }
```

File: src/cascadia/TerminalApp/TabManagement.cpp [hunk 3]
```diff
@@ -433,7 +433,13 @@ namespace winrt::TerminalApp::implementation
                         THROW_IF_FAILED(dialog->SetDefaultExtension(L"txt"));
 
                         // Default to using the tab title as the file name
-                        THROW_IF_FAILED(dialog->SetFileName((control.Title() + L".txt").c_str()));
+                        auto filename = { control.Title() };
+                        char unallowed_chars[] = "<>:\\/\"|?*";
+                        for (unsigned int i = 0; i < strlen(unallowed_chars); ++i)
+                        {
+                            filename.erase(std::remove(filename.begin(), filename.end(), unallowed_chars[i]), filename.end());
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TabManagement.cpp:440]: We're on C++20 which allows you to replace the erase-remove idiom with the more modern `std::erase`:
```cpp
std::erase(filename, unallowed_chars[i]);
```

**Comment by lhecker** [comment]
The STL has no builtin way to efficiently filter a string, so for the (theoretically) best approach you have to build your own algorithm, usually done by using a table approach (pseudo code from the top of my head):
```cpp
static constexpr std::array<bool, 128> cleanFilenameFilter{{
    0 /* NUL */, 0 /* SOH */, 0 /* STX */, 0 /* ETX */, 0 /* EOT */, 0 /* ENQ */, 0 /* ACK */, 0 /* BEL */, 0 /* BS  */, 0 /* HT  */, 0 /* LF  */, 0 /* VT  */, 0 /* FF  */, 0 /* CR  */, 0 /* SO  */, 0 /* SI  */,
    0 /* DLE */, 0 /* DC1 */, 0 /* DC2 */, 0 /* DC3 */, 0 /* DC4 */, 0 /* NAK */, 0 /* SYN */, 0 /* ETB */, 0 /* CAN */, 0 /* EM  */, 0 /* SUB */, 0 /* ESC */, 0 /* FS  */, 0 /* GS  */, 0 /* RS  */, 0 /* US  */,
    0 /* SP  */, 0 /* !   */, 1 /* "   */, 0 /* #   */, 0 /* $   */, 0 /* %   */, 0 /* &   */, 0 /* '   */, 0 /* (   */, 0 /* )   */, 1 /* *   */, 0 /* +   */, 0 /* ,   */, 0 /* -   */, 0 /* .   */, 1 /* /   */,
    0 /* 0   */, 0 /* 1   */, 0 /* 2   */, 0 /* 3   */, 0 /* 4   */, 0 /* 5   */, 0 /* 6   */, 0 /* 7   */, 0 /* 8   */, 0 /* 9   */, 1 /* :   */, 0 /* ;   */, 1 /* <   */, 0 /* =   */, 1 /* >   */, 1 /* ?   */,
    0 /* @   */, 0 /* A   */, 0 /* B   */, 0 /* C   */, 0 /* D   */, 0 /* E   */, 0 /* F   */, 0 /* G   */, 0 /* H   */, 0 /* I   */, 0 /* J   */, 0 /* K   */, 0 /* L   */, 0 /* M   */, 0 /* N   */, 0 /* O   */,
    0 /* P   */, 0 /* Q   */, 0 /* R   */, 0 /* S   */, 0 /* T   */, 0 /* U   */, 0 /* V   */, 0 /* W   */, 0 /* X   */, 0 /* Y   */, 0 /* Z   */, 0 /* [   */, 1 /* \   */, 0 /* ]   */, 0 /* ^   */, 0 /* _   */,
    0 /* `   */, 0 /* a   */, 0 /* b   */, 0 /* c   */, 0 /* d   */, 0 /* e   */, 0 /* f   */, 0 /* g   */, 0 /* h   */, 0 /* i   */, 0 /* j   */, 0 /* k   */, 0 /* l   */, 0 /* m   */, 0 /* n   */, 0 /* o   */,
    0 /* p   */, 0 /* q   */, 0 /* r   */, 0 /* s   */, 0 /* t   */, 0 /* u   */, 0 /* v   */, 0 /* w   */, 0 /* x   */, 0 /* y   */, 0 /* z   */, 0 /* {   */, 1 /* |   */, 0 /* }   */, 0 /* ~   */, 0 /* DEL */,
}};

std::wstring cleanFilename(std::wstring str) noexcept {
    std::erase_if(str, [](auto ch) {
        // This lookup is branchless: It always checks the filter, but throws
        // away the result if ch >= 128. This is faster than using `&&` (branchy).
        return cleanFilenameFilter[ch & 0x7f] & (ch < 128);
    });
    return str;
}
```

File: src/cascadia/TerminalApp/TabManagement.cpp (hunk 3 above)
Comment (EliaSchiavon) [reply src/cascadia/TerminalApp/TabManagement.cpp:440]: Maybe could be useful to put this information in the contribution guide, what do you think? So other starting developers like me know what language standard is in use and use the correct features: The development section could be a good place

**Comment by EliaSchiavon** [comment]
Wow! Thank you for all your feedback! I think that for now I'm going to correct my code with your suggestions and test it and maybe then improve it using the table. Do you think that this file is a good place to define such a table and function? Or maybe is better to put them in another file? I think that maybe the second option is better since it could be called also by other parts of the code if needed

File: src/cascadia/TerminalApp/TabManagement.cpp (hunk 3 above)
Comment (serd2011) [reply src/cascadia/TerminalApp/TabManagement.cpp:440]: You can always check C++ version in project properties.

**Comment by lhecker** [comment]
> Do you think that this file is a good place to define such a table and function?

If you choose to use the table based approach, then I'd suggest putting it into our `<til/string.h>` header: https://github.com/microsoft/terminal/blob/ffe9a0f09bc7b3cb96908d265c93bcd1a4e75fa2/src/inc/til/string.h
It already contains similar string-related functions like `visualize_control_codes`. But I'd approve this PR definitely even without it. Such a function would have no measurable benefit for this use case after all - it would just look nice (which can also be neat in a way 😅). I've adjusted my pseudo-code above to reflect the different code style in that file.

**Comment by xtremevipper** [comment]
Sorry, i am missing

File: src/cascadia/TerminalApp/TabManagement.cpp (hunk 3 above)
Comment (EliaSchiavon) [reply src/cascadia/TerminalApp/TabManagement.cpp:440]: Thank you! Didn't know about it

**Comment by EliaSchiavon** [comment]
Thank you again @lhecker ! I'm going to work on this these days

**Comment by EliaSchiavon** [comment]
I added the table in `<til/string.h>` but have not tested it yet since I had some problems with my installation of C++ with VS. I have VS2019 16.11 but when I try to build I get this error:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(439,5): error MSB8020: The build tools for v143 (Platform Toolset = 'v143') cannot be found. To build using the v143 build tools, please install v143 build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [C: \Users\eliaa\Documents\GitHub\terminal\src\tools\closetest\CloseTest.vcxproj]
```

that searching should be resolved by installing VS2022, or otherwise, the target VS should be changed by the solution properties, is this correct? Anyway, I start to send the code since the installation of VS2022 is taking some time so I can get feedback on the work

**Comment by DHowett** [comment]
Thanks for bearing with us! Yeah, unfortunately as of pretty recently we require VS 2022 😦

**Comment by lhecker** [comment]
> Thanks for bearing with us! Yeah, unfortunately as of pretty recently we require VS 2022 😦

I'll submit a PR to restore support for VS 2019 soon. We can't officially support it for this project though since only VS 2022 fully supports C++20 unfortunately.

File: src/inc/til/string.h [hunk 4]
```diff
@@ -30,6 +30,29 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
+            0 /* P   */, 0 /* Q   */, 0 /* R   */, 0 /* S   */, 0 /* T   */, 0 /* U   */, 0 /* V   */, 0 /* W   */, 0 /* X   */, 0 /* Y   */, 0 /* Z   */, 0 /* [   */, 1 /* \   */, 0 /* ]   */, 0 /* ^   */, 0 /* _   */,
+            0 /* `   */, 0 /* a   */, 0 /* b   */, 0 /* c   */, 0 /* d   */, 0 /* e   */, 0 /* f   */, 0 /* g   */, 0 /* h   */, 0 /* i   */, 0 /* j   */, 0 /* k   */, 0 /* l   */, 0 /* m   */, 0 /* n   */, 0 /* o   */,
+            0 /* p   */, 0 /* q   */, 0 /* r   */, 0 /* s   */, 0 /* t   */, 0 /* u   */, 0 /* v   */, 0 /* w   */, 0 /* x   */, 0 /* y   */, 0 /* z   */, 0 /* {   */, 1 /* |   */, 0 /* }   */, 0 /* ~   */, 0 /* DEL */,
+            // clang-format on
+        } };
+
+        std::erase_if(str, [](auto ch) {
+            // This lookup is branchless: It always checks the filter, but throws
+            // away the result if ch >= 128. This is faster than using `&&` (branchy).
+            return filter[ch & 127] & (ch < 128);
```
Comment (lhecker) [inline src/inc/til/string.h:51]: Try replacing this line with
```suggestion
            return til::at(filter, ch & 127) & (ch < 128);
```

It should fix all 4 warnings. Sorry about that!

**Review by lhecker — APPROVED** [review]

File: src/inc/til/string.h (hunk 4 above)
Comment (EliaSchiavon) [reply src/inc/til/string.h]: No, I'm sorry...I was not able to test this before sending the code 😞 I just tested the function with a string but in a small script outside the project

**Comment by EliaSchiavon** [comment]
> Edit: After some deliberation we've decided to not restore support for VS 2019. It just wouldn't work correctly.

If your decision is definitive, consider updating the readme here:

https://github.com/microsoft/terminal/blob/b55bcb50d9f0a8340e3dea9c48834ed09886c790/README.md?plain=1#L295-L296

I can do that little edit in a PR if the project will be supported on VS2022 only from now on

**Comment by lhecker** [comment]
I've opened a PR a little while ago actually: https://github.com/microsoft/terminal/pull/13709
I'm extremely sorry about that! When I upgraded the project to C++20 and VS 2022 I totally forgot to update the readme. 😥

**Review by zadjii-msft — APPROVED** [review]
Any reason this is still in draft? This looks good to me!

**Comment by EliaSchiavon** [comment]
No, forgive me! I forgot to check it as ready to merge

Il gio 11 ago 2022, 17:39 Mike Griese ***@***.***> ha scritto:

File: src/inc/til/string.h [hunk 5]
```diff
@@ -30,6 +30,29 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
+            0 /* @   */, 0 /* A   */, 0 /* B   */, 0 /* C   */, 0 /* D   */, 0 /* E   */, 0 /* F   */, 0 /* G   */, 0 /* H   */, 0 /* I   */, 0 /* J   */, 0 /* K   */, 0 /* L   */, 0 /* M   */, 0 /* N   */, 0 /* O   */,
+            0 /* P   */, 0 /* Q   */, 0 /* R   */, 0 /* S   */, 0 /* T   */, 0 /* U   */, 0 /* V   */, 0 /* W   */, 0 /* X   */, 0 /* Y   */, 0 /* Z   */, 0 /* [   */, 1 /* \   */, 0 /* ]   */, 0 /* ^   */, 0 /* _   */,
+            0 /* `   */, 0 /* a   */, 0 /* b   */, 0 /* c   */, 0 /* d   */, 0 /* e   */, 0 /* f   */, 0 /* g   */, 0 /* h   */, 0 /* i   */, 0 /* j   */, 0 /* k   */, 0 /* l   */, 0 /* m   */, 0 /* n   */, 0 /* o   */,
+            0 /* p   */, 0 /* q   */, 0 /* r   */, 0 /* s   */, 0 /* t   */, 0 /* u   */, 0 /* v   */, 0 /* w   */, 0 /* x   */, 0 /* y   */, 0 /* z   */, 0 /* {   */, 1 /* |   */, 0 /* }   */, 0 /* ~   */, 0 /* DEL */,
+            // clang-format on
+        } };
+
+        std::erase_if(str, [](auto ch) {
+            // This lookup is branchless: It always checks the filter, but throws
+            // away the result if ch >= 128. This is faster than using `&&` (branchy).
```
Comment (DHowett) [inline src/inc/til/string.h:50]: this SCREAMS "leonard code". I love it.

**Review by DHowett — APPROVED** [review]
This is excellent work, thank you for doing it!

**Comment by EliaSchiavon** [comment]
Thank you guys! Hope to work with you on other issues in the future!

**Pull request merged**